chalk-solve/src/clauses.rs

```diff
@@ -952,3 +953,109 @@ impl ToProgramClauses for AssociatedTyDatum {
         );
```

Comment:
I think we should move the `ToProgramClauses` trait and impls into a sub-module, and make it `pub(crate)` as it is never used outside this crate (or even `pub(super)`, to be more specific. The `clauses` module should probably just export a single function, like

```rust
/// Given some goal `goal` that must be proven, along with
/// its `environment`, figures out the program clauses that apply
/// to this goal from the Rust program. So for example if the goal
/// is `Implemented(T: Clone)`, then this function might return clauses
/// derived from the trait `Clone` and its impls.
pub fn program_clauses_for_goal(
    program: &dyn RustIrDatabase, // nit: I think we should rename this to `db`
    environment: &Environment, // I forget the exact name of this type
    goal: &DomainGoal,
) -> Vec<ProgramClause> {
    let mut vec = vec![];
    program_clauses_that_could_match(program, goal, &mut vec);
    program_clauses_for_env(program, environment, &mut vec);
    vec.retain(|c| c.could_match(goal));
    vec
}
```

Reply:
Makes sense!

Reply:
`ToProgramClauses` is still used here though: https://github.com/rust-lang/chalk/blob/master/src/query.rs#L126

Reply:
Oh, sorry, missed this -- yeah, we have to remove the `environment` query, and its users. We can leave it for later potentially.